server/src/main/java/org/opensearch/common/lucene/Lucene.java

```diff
@@ -162,6 +163,27 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
         return SegmentInfos.readLatestCommit(directory);
     }
 
+    /**
+     * A variant of {@link #readSegmentInfos(Directory)} that supports reading indices written by
+     * older major versions of Lucene. The underlying implementation is a workaround since the
+     * "expert" readLatestCommit API is currently package-private in Lucene. First, all commits in
+     * the given {@link Directory} are listed - this result includes older Lucene commits. Then,
+     * the latest index commit is opened via {@link DirectoryReader} by including a minimum supported
+     * Lucene major version based on the minimum compatibility of the given {@link org.opensearch.Version}.
+     */
+    public static SegmentInfos readSegmentInfosExtendedCompatibility(Directory directory, org.opensearch.Version minimumVersion)
+        throws IOException {
+        // This list is sorted from oldest to latest
+        List<IndexCommit> indexCommits = DirectoryReader.listCommits(directory);
+        IndexCommit latestCommit = indexCommits.get(indexCommits.size() - 1);
+        final int minSupportedLuceneMajor = minimumVersion.minimumIndexCompatibilityVersion().luceneVersion.major;
+        StandardDirectoryReader reader = (StandardDirectoryReader) DirectoryReader.open(latestCommit, minSupportedLuceneMajor, null);
+        // cache the SegmentInfos object before closing the reader to prevent any open file handles
+        SegmentInfos si = reader.getSegmentInfos();
```

Comment:
Wrote a (rather basic) unit test to check for this. I'd argue that the two implementations are actually coupled but it's just difficult to see given the workaround implementation in this method.

The code in `readSegmentInfos` internally invokes the "expert" API, but sets the minimum-compatible version to N-1:

https://github.com/apache/lucene/blob/3e8ef57e3fa1a015fc4ae22153660b21626886c4/lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java#L517

Ideally, the `readSegmentInfosExtendedCompatibility` method would simply invoke this expert API and override the minimum-compatible version. However, since that method signature is package-private in Lucene, the round-a-bout way to access it is through `DirectoryReader.listCommits` which uses it internally - 

https://github.com/apache/lucene/blob/3e8ef57e3fa1a015fc4ae22153660b21626886c4/lucene/core/src/java/org/apache/lucene/index/DirectoryReader.java#L267

and then enforce the minimum-compatible version via the `DirectoryReader`.

That said, I take your point that the two code flows aren't identical and could change underneath us, so the unit test is valuable.